dali/pipeline/data/tensor_vector.cc

```diff
@@ -507,26 +658,62 @@ void TensorVector<Backend>::Reset() {
 template <typename Backend>
 template <typename SrcBackend>
 void TensorVector<Backend>::Copy(const TensorList<SrcBackend> &in_tl, AccessOrder order) {
+  // This variant will be removed with the removal of TensorList.
   SetContiguous(true);
+
+  auto copy_order = copy_impl::SyncBefore(this->order(), in_tl.order(), order);
+
+
+  tl_->Resize(in_tl.shape(), in_tl.type());
+  type_ = in_tl.type_info();
+  sample_dim_ = in_tl.shape().sample_dim();
+
+  copy_impl::SyncAfterResize(this->order(), copy_order);
+
+  // Update the metadata
   type_ = in_tl.type_info();
   sample_dim_ = in_tl.shape().sample_dim();
-  tl_->Copy(in_tl, order);
+
+  // Here both batches are contiguous
+  type_info().template Copy<Backend, SrcBackend>(
+      unsafe_raw_mutable_data(*tl_), unsafe_raw_data(in_tl), in_tl.shape().num_elements(),
+      copy_order.stream(), false);
+  copy_impl::SyncAfter(this->order(), copy_order);
 
   resize_tensors(tl_->num_samples());
   UpdateViews();
+
+  // Update the layout and other metadata
+  SetLayout(in_tl.GetLayout());
+  for (int i = 0; i < curr_num_tensors_; i++) {
+    SetMeta(i, in_tl.GetMeta(i));
+  }
```

Comment:
Doesn't UpdateViews do it?

Reply:
Not really, it just copies over the metadata from internal tl_ to the tensors_.

Reply:
So, it _does_ it, and, again, you just want to pretend that tl_ doesn't exist....?

Reply:
Ok, whatever, I will revert this to the previous variant and remove it in the next PR.